### PR TITLE
added support for custom headers

### PR DESCRIPTION
**Pull request by macnev2013**
**Issues**
- [4133](https://github.com/localstack/localstack/issues/4133)

**Added**
- Added support for custom headers in s3

### CODE REVIEW

File: localstack/services/s3/s3_utils.py [hunk 1]
```diff
@@ -192,7 +192,7 @@ def authenticate_presign_url(method, path, headers, data=None):
                     elif key_lower in (blacklisted_header.lower() for blacklisted_header in BLACKLISTED_HEADERS):
                         pass
                     else:
-                        sign_headers[key] = query_params[key][0]
+                        query_string[key] = query_params[key][0]
```
Comment (cmlara) [inline localstack/services/s3/s3_utils.py:195]: I had an issue in my lab when I tried this. Several tests that are using the V2 signature started failing.  Maybe it was just my lab but reverting that single change caused the tests to start working.

**Comment by cmlara** [comment]
Can I suggest a test is added? 

Here is what I have been working with in my lab off and on trying to debug this.

```
diff --git a/tests/integration/test_s3.py b/tests/integration/test_s3.py
index d9093c84..30c881a1 100644
--- a/tests/integration/test_s3.py
+++ b/tests/integration/test_s3.py
@@ -2280,6 +2280,36 @@ class TestS3(unittest.TestCase):
             result = getattr(requests, method.lower())(url, data=body, headers=headers)
             self.assertLess(result.status_code, 400)
 
+    def add_query_param(self, request, **kwargs):
+        request.url += 'requestedBy=abcDEF123'
+
+    def test_presign_with_query_params(self):
+
+        bucket_name = short_uid()
+        s3_client = aws_stack.connect_to_service('s3')
+
+        s3_presign = boto3.client('s3',
+            endpoint_url=config.get_edge_url(),
+            aws_access_key_id='test',
+            aws_secret_access_key='test',
+            config=Config(signature_version='s3v4')
+        )
+
+        s3_client.create_bucket(Bucket=bucket_name)
+        s3_client.put_object(Body='test-value', Bucket=bucket_name, Key='test')
+        response = s3_client.head_object(Bucket=bucket_name, Key='test')
+
+        s3_client.meta.events.register('before-sign.s3.GetObject', self.add_query_param)
+
+        presign_url = s3_presign.generate_presigned_url(
+            ClientMethod='get_object',
+            Params={'Bucket': bucket_name, 'Key': 'test'},
+            ExpiresIn=86400
+        )
+
+        response = requests.get(presign_url)
+        self.assertEqual(response._content, b'test-value')
+
     # ---------------
     # HELPER METHODS
     # ---------------

```

**Comment by cmlara** [comment]
Disregard my test code suggestion. It looks like when I got distracted seeing V2 tests failures I forgot to verify my own test correctly failed without the patch (it doesn't)

**Comment by macnev2013** [comment]
Thanks for the suggestion @cmlara . We have added the test case you have provided. Thanks.

**Comment by macnev2013** [comment]
Looks like the added test cases was somehow interfering with other test cases. I have removed it for now. I'll add it in follow up PR.

File: tests/integration/test_s3.py [hunk 2]
```diff
@@ -1625,6 +1625,33 @@ def test_presigned_url_signature_authentication(self):
...
+            endpoint_url=config.get_edge_url(),
+            aws_access_key_id='test',
+            aws_secret_access_key='test',
+            config=Config(signature_version='s3v4')
+        )
+
+        s3_client.create_bucket(Bucket=bucket_name)
+        s3_client.put_object(Body='test-value', Bucket=bucket_name, Key='test')
+        response = s3_client.head_object(Bucket=bucket_name, Key='test')
+        s3_client.meta.events.register('before-sign.s3.GetObject', self.add_query_param)
```
Comment (whummer) [inline tests/integration/test_s3.py:1644]: nit: let's better wrap this code in a `try-finally` block (running the `unregister(..)` below in the `finally` block), to ensure that the event handler is unregistered even if this tests raises an error.

Also, let's add a short note nere that this test may have side effects (and hence may be problematic for parallelization):
```
# Note: This test may have side effects (via `s3_client.meta.events.register(..)`) and may not be suitable for parallel execution
```

Comment (macnev2013) [reply tests/integration/test_s3.py]: Thanks for the suggestions. I have updated the code accordingly.

**Comment by whummer** [comment]
Thanks!

**Pull request merged**